cumulus_library/studies/core/builder_encounter_coding.py

```diff
@@ -0,0 +1,123 @@
+                    query = get_is_table_not_empty_query(
+                        "encounter",
+                        "t1.row1",
+                        [
+                            {
+                                "source_col": code_source["name"],
+                                "table_alias": "t1",
+                                "row_alias": "row1",
+                            }
+                        ],
+                    )
+                else:
+                    query = get_is_table_not_empty_query(
+                        "encounter", f"{code_source['name']}.coding"
+                    )
+                cursor.execute(query)
+                progress.advance(task)
+
+                if len(cursor.fetchall()) > 0:
+                    if code_source["is_array"]:
+                        query = get_is_table_not_empty_query(
+                            "encounter",
+                            "t2.row2",
+                            [
+                                {
+                                    "source_col": code_source["name"],
+                                    "table_alias": "t1",
+                                    "row_alias": "row1",
+                                },
+                                {
+                                    "source_col": "row1.coding",
+                                    "table_alias": "t2",
+                                    "row_alias": "row2",
+                                },
+                            ],
+                        )
+                    else:
+                        query = get_is_table_not_empty_query(
+                            "encounter",
+                            "t1.row1",
+                            [
+                                {
+                                    "source_col": f"{code_source['name']}.coding",
+                                    "table_alias": "t1",
+                                    "row_alias": "row1",
+                                }
+                            ],
+                        )
+                    cursor.execute(query)
+                    progress.advance(task)
+                    if len(cursor.fetchall()) > 0:
+                        cols_with_data.append(code_source)
+                else:
+                    progress.advance(task)
+            else:
+                progress.advance(task)
+                progress.advance(task)
+        return cols_with_data
+
+    def prepare_queries(self, cursor: object, schema: str):
+        """Constructs queries related to condition codeableConcept
+
+        :param cursor: A database cursor object
+        :param schema: the schema/db name, matching the cursor
+
+        """
+        code_sources = [
+            {"name": "type", "is_array": True},
+            {"name": "servicetype", "is_array": False},
+            {"name": "priority", "is_array": False},
+        ]
+        discovery_length = len(code_sources) * 3
```

Comment:
I don't love this construction - a magic value 3 that requires this line to know a lot about the internals of a separate method.

What if you pass `total=None` when constructing the task, and when advancing, you can just call `progress.update(task, completed=2, total=3)` (which also lets you skip double-advance calls by just setting `completed` or `total` directly).

Reply:
it was more sensible (alebit less readable) before i pulled the ladder logic out into its own function.

but my thought was this:
- you will always need to run three queries to prove conclusively that data exists, without introducing intentional error catching for misfired queries, which I think we don't want to do because it masks bug behavior
- We want someone to be able to add/remove codings from this easily without thinking too hard about it.

I could do the `total=None` route, though I selfishly like an advancing one for communicating that this is a known length of stuff we're doing here. I haven't used this in this way before so i don't fully know what the UX looks like.

what if 3 was replaced by a semantically named constant?

Reply:
My beef is less with wanting a semantically named magic constant, and more with the separation of logic between two different methods. If the `* 3` were at the top of `_check_data_in_fields` (like maybe `progress.update(task, total=len()*3)`?), I'd be happy enough. I just want that method boundary to actually be a semantic boundary, and allow the method to be rewritten in the future.

I think if the `* 3` is close to the code that does something thrice and there's a little comment, I don't think the addition of a named constant would add much more.

Reply:
Or heck, have the method own & make its own task?

Reply:
Ok, i went ahead and pushed the task up into this method, so we've got a full boundary between 'check if data exists' and 'build a table with columns that have data'